src/sidebar/components/search-input.js

```diff
@@ -1,45 +1,92 @@
 'use strict';
 
-// @ngInject
-function SearchInputController($element, store) {
-  const self = this;
-  const button = $element.find('button');
-  const input = $element.find('input')[0];
-  const form = $element.find('form')[0];
-
-  button.on('click', function() {
-    input.focus();
-  });
-
-  form.onsubmit = function(e) {
-    e.preventDefault();
-    self.onSearch({ $query: input.value });
-  };
+const classnames = require('classnames');
+const { createElement } = require('preact');
+const { useRef, useState } = require('preact/hooks');
+const propTypes = require('prop-types');
 
-  this.isLoading = () => store.isLoading();
+const useStore = require('../store/use-store');
 
-  this.inputClasses = function() {
-    return { 'is-expanded': self.alwaysExpanded || self.query };
-  };
+const Spinner = require('./spinner');
+
+/**
+ * An input field in the top bar for entering a query that filters annotations
+ * (in the sidebar) or searches annotations (in the stream/single annotation
+ * view).
+ *
+ * This component also renders a loading spinner to indicate when the client
+ * is fetching for data from the API or in a "loading" state for any other
+ * reason.
+ */
+function SearchInput({ alwaysExpanded, query, onSearch }) {
+  const isLoading = useStore(store => store.isLoading());
+  const input = useRef();
+
+  // The active filter query from the previous render.
+  const [prevQuery, setPrevQuery] = useState(query);
+
+  // The query that the user is currently typing, but may not yet have applied.
+  const [pendingQuery, setPendingQuery] = useState(query);
 
-  this.$onChanges = function(changes) {
-    if (changes.query) {
-      input.value = changes.query.currentValue;
-    }
+  const onSubmit = e => {
+    e.preventDefault();
+    // TODO - When the parent components are converted to React, the signature
+    // of the callback can be simplified to `onSearch(query)` rather than
+    // `onSearch({ $query: query })`.
+    onSearch({ $query: input.current.value });
   };
+
+  // When the active query changes outside of this component, update the input
+  // field to match. This happens when clearing the current filter for example.
+  if (query !== prevQuery) {
+    setPendingQuery(query);
+    setPrevQuery(query);
+  }
+
+  return (
+    <form className="search-input__form" name="searchForm" onSubmit={onSubmit}>
+      <input
+        className={classnames('search-input__input', {
+          'is-expanded': alwaysExpanded || query,
+        })}
+        type="text"
+        name="query"
+        placeholder={(isLoading && 'Loading…') || 'Search…'}
+        disabled={isLoading}
+        ref={input}
+        value={pendingQuery}
+        onInput={e => setPendingQuery(e.target.value)}
```

Comment:
I feel naïve here. Given that `setPendingQuery` will update state, is the reality that this component will re-render on every keystroke (input event)?

Reply:
Yes it will, and that's fine. What is a problem is if your app has a large component tree and re-renders on every keystroke or otherwise triggers any work which a significant amount of time to run.